### PR TITLE
full support to PHP 8.0 + add support for CurlHandle resource objects

### LINKED ISSUE #68: PHP 8 Support
This will require some work, since the curl functions no longer return resources.

### PR DESCRIPTION
**Pull request by ildyria**
| Q               | A
| --------------- | ---
| Bug fix?         | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes #68, closes #70
| Documentation   | -
| License         | MIT

#### What's in this PR?

- In PHP 8.0, [Curl resources are changed to CurlHandle class objects](https://php.watch/versions/8.0/resource-CurlHandle#is-resource). This changes DocBlock parameters and other areas where provided handles are checked with `is_resource()`, which now needs to account for the replaced `CurlHandle` class objects.

- In order to apply the test, it uses the github workflows.

- Migrate from php-http/client-integration-tests from `^2.0` to `^3.0`

#### Why?

- The author of #70 has not reacted since his PR, so nothing is moving.

- Because in PHP 8.0, Curl handles are no longer resources, and this libraries `is_resource` no longer return true for valid Curl handlers.

- Migrate from php-http/client-integration-tests so that PHP 8.0 is supported.

#### Example Usage

``` php
is_resource(curl_init()); // true in PHP < 8.0
is_resource(curl_init()); // false in PHP >= 8.0
```

### CODE REVIEW

**Comment by ildyria** [comment]
@dbu I incorporated your suggestions to #70 in this PR.

**Comment by ildyria** [comment]
Full CI run here: https://github.com/ildyria/curl-client/actions/runs/413449950

**Review by dbu — COMMENTED** [review]
thanks!

can you delete the .travis.yml file as well? when you move the build to github actions, we don't need travis anymore.

**Comment by ildyria** [comment]
@dbu fixed as per requested.

File: .github/workflows/Build-Test.yml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+jobs:
+  tests:
+    if: "! contains(toJSON(github.event.commits.*.msg), 'skip') && ! contains(toJSON(github.event.commits.*.msg), 'ci')" #skip ci...
+    runs-on: ${{ matrix.operating-system }}
+
+    strategy:
+      fail-fast: false
+      matrix:
+        operating-system: [ubuntu-16.04,ubuntu-18.04,ubuntu-20.04]
```
Comment (dbu) [inline .github/workflows/Build-Test.yml:14]: sorry to only notice now, but why build on all ubuntu LTS versions? on travis-ci we only built with whatever the default environment was. i would suggest that ubuntu-20.04 is enough. what we do care about is the various php minor versions, but we don't expect changes between ubuntu versions. that would reduce the number of builds from 13 to 5 if i understand correctly.

Comment (ildyria) [reply .github/workflows/Build-Test.yml]: fixed with latest commit.

**Comment by dbu** [comment]
thanks a lot, this is coming good!

as you expected, the coverage fails in your repository. i will merge and see if it works from our repository.

File: .github/workflows/Build-Test.yml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+      - name: Cache dependencies
+        uses: actions/cache@v2
+        with:
+          path: ${{ steps.composer-cache.outputs.dir }}
+          key: ${{ runner.os }}-composer-${{ hashFiles('**/composer.lock') }}
+          restore-keys: ${{ runner.os }}-composer-
+
+      - name: Composer self update
+        run: composer self-update
```
Comment (GrahamCampbell) [inline .github/workflows/Build-Test.yml:75]: This step should definitely not be here.

File: .github/workflows/Build-Test.yml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+      - name: Install Composer dependencies
+        run: composer update ${{ matrix.COMPOSER_FLAGS }} --prefer-source --no-interaction
```
Comment (GrahamCampbell) [inline .github/workflows/Build-Test.yml:78]: prefer-source will be very slow

File: .github/workflows/Build-Test.yml [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+      - name: boot test server
+        run: vendor/bin/http_test_server > /dev/null 2>&1 &
+
+      - name: Apply tests
```
Comment (GrahamCampbell) [inline .github/workflows/Build-Test.yml:83]: run?

**Comment by dbu** [comment]
thanks for the review graham - i was so excited for php 8 that i missed those details. i do a cleanup in #74

**Comment by Ayesh** [comment]
Author of #70 - thank you for picking up it in here, I'm happy to see it merged now. I apologize about my delays in the other PR, I was having some issues lately. 🙏🏼

**Comment by dbu** [comment]
thanks for having gotten the ball rolling @Ayesh and @ildyria for wrapping it up. collaboration for the win!

**Pull request merged**